lua/dap/repl.lua

```diff
@@ -197,7 +214,45 @@ local function print_threads(threads)
 end
 
 
+---- Handle multiline inputs
+--
+-- Returns nil if input is not complete, otherwise returns the full user input
+local function handle_followup_input(curr_text)
+  local session = get_session()
+  if not session then
+    return curr_text
+  end
+  local continue_prompt_fn = session.adapter.followup_func_cb
+  if not continue_prompt_fn then
+    return curr_text
+  end
+  local prompt_state = vim.fn.getbufvar(repl.buf, PROMPT_STATE)
+  if prompt_state == REGULAR_PROMPT and continue_prompt_fn(curr_text) then
+    vim.fn.setbufvar(repl.buf, PROMPT_STATE, FOLLOWUP_PROMPT)
+    vim.fn.setbufvar(repl.buf, 'curr_text', curr_text .. '\n')
```

Comment:
Here, if you add the text as it comes, you will add the text with the `\` character at the end. This probably works in python, but it doesn't work in other languages, like javascript. So, I think it would be nicer if the `followup_func_cb` would return the text to actually be processed, or nil if it considers that it's not a multi line prompt. Then we could be developing something like this: 

```lua
followup_func_cb = function(text)
    if string.sub(text, -1) == '\\' then
        return string.gsub(text, "\\[ ]*$", "")
    end
    return nil
end
```

for `javascript`, or something like this, for python:

```lua
followup_func_cb = function(text)
  if string.sub(text, -1) == ':' then
    return text
  elseif string.sub(text, -1) == '\\' then
    return text
  elseif select(2, string.gsub(text, '"""', "")) == 1 then
    return text
  end
  return nil
end
```

This would give a much larger flexibility, since would basically work for any language, not just for python.

Of course, the `repl.lua` file should be modified a little: 

```patch
diff --git a/lua/dap/repl.lua b/lua/dap/repl.lua
index ac8d21e..1d8a83c 100644
--- a/lua/dap/repl.lua
+++ b/lua/dap/repl.lua
@@ -227,20 +227,21 @@ local function handle_followup_input(curr_text)
     return curr_text
   end
   local prompt_state = vim.fn.getbufvar(repl.buf, PROMPT_STATE)
-  if prompt_state == REGULAR_PROMPT and continue_prompt_fn(curr_text) then
+  local txt = continue_prompt_fn(curr_text)
+  if prompt_state == REGULAR_PROMPT and txt ~= nil then
     vim.fn.setbufvar(repl.buf, PROMPT_STATE, FOLLOWUP_PROMPT)
-    vim.fn.setbufvar(repl.buf, 'curr_text', curr_text .. '\n')
+    vim.fn.setbufvar(repl.buf, 'curr_text', txt .. '\n')
     vim.fn.prompt_setprompt(repl.buf, get_prompt())
     return nil
   elseif prompt_state == FOLLOWUP_PROMPT then
     local previous_text = vim.fn.getbufvar(repl.buf, 'curr_text')
-    if curr_text == '' then
+    if txt == nil then
       curr_text = previous_text .. curr_text
       vim.fn.setbufvar(repl.buf, PROMPT_STATE, REGULAR_PROMPT)
       vim.fn.prompt_setprompt(repl.buf, get_prompt())
       return curr_text
     else
-      vim.fn.setbufvar(repl.buf, 'curr_text', previous_text .. curr_text .. '\n')
+      vim.fn.setbufvar(repl.buf, 'curr_text', previous_text .. txt .. '\n')
       return nil
     end
   end
```

Reply:
I am a bit hesitant to add such a feature since it wouldn't be obvious to the user what was the exact input sent to the debug adapter. But if the purpose of this logic is to only start multi-line mode, then we could add Shift-Enter binding that will enter multi-line mode and go to new line, this could be active without needing any additional configuration. Another possibility for the user is to use treesitter in `followup_func_cb` and if the line gives an error or in python's case is a class or function definition then return true.
 @cosminadrianpopescu what do you think?

Reply:
I understand the concern. But the way is implemented right now, it only works for python.

> then we could add Shift-Enter binding that will enter multi-line mode and go to new line, this could be active without needing any additional configuration

This sounds good, but it will only work in certain terminals. `Shift + Enter` is not very easy to set up in vim, due to terminal restrictions (tty is even more difficult).

> Another possibility for the user is to use treesitter in followup_func_cb and if the line gives an error or in python's case is a class or function definition then return true.

This would not really help, since the text would still contain the `\` character at the end, so still would not work in other languages. 

Anoter possibility is to have a setting, which would indicate if the character should be added to the query or not. Like `g:dap_repl_multiline_ignore`. Something on those lines, if clarity is an issue. 

However, in my oppinion, since the `follow_func_cb` is not mandatory, who will want to use it, will not have issues understanding the concept.

Reply:
Here is an example of how to use treesitter for this:
```lua
local string_queries = { '(ERROR) @capture', '(expression_statement (function) @capture)' }
local queries = {}
for _, query in ipairs(string_queries) do
  table.insert(queries, vim.treesitter.query.parse_query('javascript', query))
end

local function followup_func_cb(text)
      local parser = vim.treesitter.get_string_parser(text, 'javascript')
      local tree = parser:parse()[1]
      local results = {}
      for _, query in ipairs(queries) do
        for id, _, _ in query:iter_captures(tree:root(), text) do
          table.insert(results, id)
        end
      end
      if #results > 0 then
        return true
      end
      return false
end
```
Here are the results:
![image](https://user-images.githubusercontent.com/20954878/208686343-1670112d-cb32-4530-bfd6-82298ce36cac.png)
I think that in general the best way to check if a follow up prompt should appear is to use treesitter or with more complex logic rather than just looking at the last letter. 

> However, in my oppinion, since the follow_func_cb is not mandatory, who will want to use it, will not have issues understanding the concept.

I assume most people who use dap are not configuring it by themselves, either using a plugins that already provide the configuration or because they are using a neovim distribution.

If you are not convinced it's good enough let me know and I'll make the change.